python/ray/data/grouped_data.py

```diff
@@ -203,47 +187,29 @@ def map_groups(
         else:
             sorted_ds = self._dataset.repartition(1)
 
-        def get_key_boundaries(block_accessor: BlockAccessor) -> List[int]:
-            """Compute block boundaries based on the key(s)"""
-
-            import numpy as np
-
-            # Get the keys of the batch in numpy array format
-            keys = block_accessor.to_numpy(self._key)
-
-            if isinstance(keys, dict):
-                # For multiple keys, we generate a separate tuple column
-                convert_to_multi_column_sorted_key = np.vectorize(_MultiColumnSortedKey)
-                keys: np.ndarray = convert_to_multi_column_sorted_key(*keys.values())
-
-            boundaries = []
-            start = 0
-            while start < keys.size:
-                end = start + np.searchsorted(keys[start:], keys[start], side="right")
-                boundaries.append(end)
-                start = end
-            return boundaries
-
         # The batch is the entire block, because we have batch_size=None for
         # map_batches() below.
         def apply_udf_to_groups(udf, batch, *args, **kwargs):
             block = BlockAccessor.batch_to_block(batch)
             block_accessor = BlockAccessor.for_block(block)
+
+            # Get the list of boundaries including first start and last end indices
             if self._key:
-                boundaries = get_key_boundaries(block_accessor)
+                # Get a block of key columns
+                projected_block = block_accessor.to_numpy(self._key)
```

Comment:
So the only case when `projected_block` will be just `ndarray` is when the `_key` is a single tensor column (which doesn't really make sense by itself but that's a separate story).

Let's just convert it back to a dict like following:

```
if isinstance(projected_block, np.ndarray):
  projected_block = {TENSOR_COLUMN_NAME: projected_block}
```